File: 2022/10/22/22.ts
(() => {
    class Node {
        constructor(
            public val: string,
            public left: Node | null = null,
            public right: Node | null = null
        ) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    const serialize = (node: Node | null): string => {
        if (!node) return 'X';
        return `${node.val} ${serialize(node.left)} ${serialize(node.right)}`;
    };

    const deserialize = (serialized: string): Node | null => {
        const nodes = serialized.split(' ');
        const buildNode = (): Node | null => {
            const val = nodes.shift();
            if (val === 'X') return null;
            return new Node(val as string, buildNode(), buildNode());
        };
        return buildNode();
    }

    const node = new Node('root', new Node('left', new Node('left.left')), new Node('right'));
    console.log(serialize(node));
    console.log(deserialize(serialize(node)));
    console.log(node.left?.left?.val);
    if (deserialize(serialize(node))?.left?.left?.val === 'left.left') {
        console.log('Deserialized correctly');
    } else {
        console.log('Deserialized incorrectly');
    }
})();

File: 2022/10/22/22.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def serialize(node: Node) -> str:
    if node is None:
        return 'X'
    return f"{node.val} {serialize(node.left)} {serialize(node.right)}"


def deserialize(serialized: str) -> Node or None:
    nodes = serialized.split()

    def generate_node() -> Node or None:
        val = nodes.pop(0)
        if val == 'X':
            return None
        return Node(val, generate_node(), generate_node())

    return generate_node()


node = Node('root', Node('left', Node('left.left')), Node('right'))
print(serialize(node))
assert deserialize(serialize(node)).left.left.val == 'left.left'
assert node.left.left.val == 'left.left'
assert (deserialize(serialize(node)).right.val == 'right')
assert node.right.val == 'right'
